misc/mcp_server_stateless.py

```diff
@@ -0,0 +1,108 @@
+# ---
+# cmd: ["modal", "serve", "misc/mcp_server_stateless.py"]
+# ---
+#
+# # Deploy a remote, stateless MCP server on Modal
+#
+# This example demonstrates how to deploy a simple [MCP
+# server](https://modelcontextprotocol.io/) on Modal, that provides a tool to get the
+# current date and time in a given timezone. It is a stateless MCP server, meaning that
+# it does not store any state between requests, and uses the "streamable HTTP" transport
+# type.
+#
+# It uses the [FastMCP library](https://github.com/jlowin/fastmcp) to create the MCP
+# server, which is wrapped using FastAPI, and served with a [Modal web
+# endpoint](https://modal.com/docs/guide/webhooks).
+
+
+import modal
+
+app = modal.App("datetime-mcp-server")
+
+image = modal.Image.debian_slim(python_version="3.12").pip_install(
+    "fastapi==0.115.14",
+    "fastmcp~=2.10.2",
+)
+
+
+# First, we create the MCP server itself using FastMCP, and add a tool to it that
+# allows LLMs to get the current date and time in a given timezone.
+def make_mcp_server():
+    from fastmcp import FastMCP
+
+    mcp = FastMCP("Date and Time MCP Server")
+
+    @mcp.tool()
+    async def current_date_and_time(timezone: str = "UTC") -> str:
+        """Get the current date and time.
+
+        Args:
+            timezone: The timezone to get the date and time in (optional). Defaults to UTC.
+
+        Returns:
+            The current date and time in the given timezone, in ISO 8601 format.
+        """
+        from datetime import datetime
+        from zoneinfo import ZoneInfo
+
+        try:
+            tz = ZoneInfo(timezone)
+        except Exception:
+            raise ValueError(
+                f"Invalid timezone '{timezone}'. Please use a valid timezone like 'UTC', "
+                "'America/New_York', or 'Europe/Stockholm'."
+            )
+        return datetime.now(tz).isoformat()
+
+    return mcp
+
+
+# We then use FastMCP to create a Starlette app with `streamable-http` as transport
+# type, and set `stateless_http=True` to make it stateless. This will be mounted by
+# FastAPI.
+def make_fastapi_app():
+    from fastapi import FastAPI
+
+    mcp = make_mcp_server()
+    mcp_app = mcp.http_app(transport="streamable-http", stateless_http=True)
+
+    fastapi_app = FastAPI(lifespan=mcp_app.router.lifespan_context)
+    fastapi_app.mount("/", mcp_app, "mcp")
+
+    return fastapi_app
+
+
+# Finally we deploy the FastAPI app as a Modal web endpoint, using [the `asgi_app`
+# decorator](https://modal.com/docs/reference/modal.asgi_app):
+@app.function(image=image)
+@modal.asgi_app()
+def web_endpoint():
```

Comment:
Hmm, why not just make this function return the FastAPI app?

IMO splitting them into two functions is more confusing than clarifying. We want to express to users that they can just take whatever function previously built their FastAPI app and wrap it with `asgi_app` to deploy to Modal.